library/core/src/intrinsics.rs

```diff
@@ -712,8 +712,12 @@ extern "rust-intrinsic" {
 
     /// Aborts the execution of the process.
     ///
-    /// A more user-friendly and stable version of this operation is
-    /// [`std::process::abort`](../../std/process/fn.abort.html).
+    /// [`std::process::abort`](../../std/process/fn.abort.html) is to be preferred if possible,
+    /// as its behaviour is more user-friendly and more stable.
+    ///
+    /// The current implementation of `intrinsics::abort` (ab)uses a debug trap.  On Unix, the
```

Comment:
I would not call it "debug trap" -- it's `llvm.trap` and there is also a `llvm.debugtrap` intrinsic, which is different.

Reply:
According to a local llvm expert[1]  in pers comm the difference between `llvm.trap` and `lvm.debugtrap` is that `trap` drops you into the debugger but is not intended to be continueable from, whereas `debugtrap` is.  Whether this is true may be doubtful, and it's definitely at the very least is not entirely clear from the documentation:
 * https://llvm.org/docs/LangRef.html#llvm-trap-intrinsic
 * https://llvm.org/docs/LangRef.html#llvm-debugtrap-intrinsic

[1]  Not mdw, who co-wrote some of the discussion of `abort(3)`.

However, it *does* say that `trap` turns into a "trap instruction".  In the general case there is no *instruction* that will perform an abort the same way as `abort()`.

Empirically some platforms `llvm.trap` does indeed result in `SIGTRAP`, which is the signal for a debug trap.  So on those platforms a debug trap is indeed being abused.

How about "(ab)uses a debug trap on some popular platforms"?  I am triying to avoid discussing llvm intrinsics here while producing the appropriate amount of FUD about the effect of `intrinsics::abort` :-).

Reply:
> popular platforms

I think x86 is a popular platform, but it produces `SIGILL`, which does not look like a debug trap to me...

Reply:
> > popular platforms
> 
> I think x86 is a popular platform, but it produces `SIGILL`, which does not look like a debug trap to me...

Yes.  "*some* popular platforms".  I think ARM64 counts :-).

> Also there are some discussions about it in #81895 and #40230.

Thanks for the references!  I should have done my own search.  I particularly agree with

> Related: I'd argue `core::intrinsics::abort()` should also be renamed `core::intrinsics::trap()`, which reflects better what it does, matches the usual lingo and makes it less prone to be confused with the usual `std::process::abort()`, etc. For similar reasons I sent #81530 in an attempt to avoid potential confusion with an `abort` vs. `unreachable` instance in wasm.

from https://github.com/rust-lang/rust/issues/81895#issuecomment-783153533

but let's do one thing at a time.

Reply:
> Yes. "_some_ popular platforms".

Oh, "some" is important... :sweat_smile: